flake.nix

```diff
@@ -11,60 +11,96 @@
-                  config.hsPkgs.markdown-unlit
-                ];
-                plutarch-test.components.exes.plutarch-test.build-tools = [
-                  config.hsPkgs.hspec-discover
+        inputs.hercules-ci-effects.flakeModule
+        inputs.pre-commit-hooks.flakeModule
+      ];
+      systems = [ "x86_64-linux" "aarch64-darwin" "x86_64-darwin" "aarch64-linux" ];
+      herculesCI = {
+        ciSystems = [ "x86_64-linux" ];
+        onPush.default.outputs = self.checks.x86_64-linux;
+      };
+      perSystem = { config, system, ... }:
+        let
+          pkgs =
+            import nixpkgs {
+              inherit system;
+              overlays = [
+                haskell-nix.overlay
+                iohk-nix.overlays.crypto
+              ];
+              inherit (haskell-nix) config;
+            };
+          project = pkgs.haskell-nix.cabalProject' {
+            src = ./.;
+            compiler-nix-name = "ghc963";
+            inputMap = {
+              "https://input-output-hk.github.io/cardano-haskell-packages" = CHaP;
+            };
+            shell = {
+              withHoogle = true;
+              exactDeps = false;
+              nativeBuildInputs = [
+                project.hsPkgs.hspec-discover.components.exes."hspec-discover"
+                project.hsPkgs.markdown-unlit.components.exes."markdown-unlit"
+              ];
+              shellHook = config.pre-commit.installationScript;
+            };
+          };
+          flake = project.flake { };
+        in
+        {
+          pre-commit = {
+            settings = {
+              src = ./.;
+              settings = {
+                ormolu.defaultExtensions = [
+                  "TypeApplications"
+                  "PatternSynonyms"
                 ];
               };
-            })
-          ];
-        })
-      ];
-
-      systems =
-        if builtins.hasAttr "currentSystem" builtins
-        then [ builtins.currentSystem ]
-        else [ "x86_64-linux" "aarch64-linux" ];
 
-      hercules-ci.github-pages.branch = "master";
-      herculesCI.ciSystems = [ "x86_64-linux" ];
-
-      perSystem = { config, pkgs, self', system, ... }: {
-        packages.combined-docs = pkgs.runCommand "combined-docs" { buildInputs = with config.packages; [ docs haddock ]; } ''
-          mkdir -p $out/share/doc
-          cp -r ${config.packages.docs}/* $out
-          cp -r ${config.packages.haddock}/share/doc/* $out/share/doc
-        '';
-        hercules-ci.github-pages.settings.contents = config.packages.combined-docs;
+              hooks = {
+                nixpkgs-fmt.enable = true;
+                cabal-fmt.enable = true;
+                fourmolu = {
+                  enable = true;
+                  excludes = [ "\.lhs" ];
+                };
+                hlint.enable = false;
+                statix.enable = true;
+                deadnix.enable = true;
+              };
```

Comment:
I would rather extract the pre-commit stuff into `pre-commit.nix` to keep the flake.nix smaller as I did https://github.com/mlabs-haskell/lambdabuffers-cardano-demo/blob/main/pre-commit.nix.

Then you just add an import as https://github.com/mlabs-haskell/lambdabuffers-cardano-demo/blob/20e43a95ad5f8927985743187ab15d2648bebbfe/flake.nix#L17

Honestly, I would also extract Hercules stuff into hercules-ci.nix.

Reply:
For small nix files like what we have here, I prefer to keep things together so someone else can get better idea of what's going on without reading a gazillion file. Especially here since Plutarch only uses Haskell, and, therefore, we won't really add much more to the current flake file